### PR TITLE
Alexdb/better structural sharing

### LINKED ISSUE #6306: feat: Better structural sharing in tanstack-query
### Describe the feature you'd like to request

TRPC uses `superjson` by default to provide types beyond json-serializable ones, however this breaks default structural sharing in tanstack-query.
The most obvious impact is that any native `Date` field in a response object will cause the data object returned by a `useQuery` hook to have a different identity.

Discussion for context: https://github.com/TanStack/query/pull/6481

### Describe the solution you'd like to see

Given that trpc uses `superjson` by default, I think I'd like to see a default `structuralSharing` function that handles equality for the types handled by superjson.

### Describe alternate solutions

* Userland functions
* An external package that provides the functionality and can be easily dropped in. This might already exist but I was not able to find a suitable lib.
* Lobby popular parsing packages (superjson, I use devalue) to add `replaceEqualDeep` equivalent functions, since the logic of the two operations are tightly related. Using two different algos could introduce very hard-to-detect bugs.

### Additional information

Adding a default function would be a breaking change :)

I don't know what the performance impact of everything would be.

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

### PR DESCRIPTION
**Pull request by aldebout**
Closes #6306

## 🎯 Changes

This PR introduces a new function to handle structural sharing of non-JSON-serializable results in tanstack-query.
With this feature, API endpoints that returns `Date`s, `Set`s and `Map`s should now have more stable data object identity between renders.

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Comment by aldebout** [comment]
Current state:

* The [first commit](https://github.com/trpc/trpc/pull/6311/commits/b59bf5b7b47cda96dda506ea6e8d0d23c85b17ca) introduces the function and exports it, the [second commit](https://github.com/trpc/trpc/pull/6311/commits/71e835dc57110ab3e10d694def8c9725c5a03e9e) uses it by default making this a breaking change.
* I would appreciate some help writing tests for the second part, I'm not familiar with testing hooks in components, we need to get the first result, invalidate, get the second result and compare both.
* I'm not sure where we should reflect this in the docs?

File: packages/react-query/src/utils/structuralSharing.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import equal from 'fast-deep-equal/es6';
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:1]: We don't like to have dependencies, so please remove this

File: packages/react-query/package.json [hunk 2]
```diff
@@ -94,5 +94,8 @@
   },
   "funding": [
     "https://trpc.io/sponsor"
-  ]
+  ],
+  "dependencies": {
+    "fast-deep-equal": "^3.1.3"
```
Comment (KATT) [inline packages/react-query/package.json:99]: We don't want dependencies

```suggestion
```

File: packages/tests/server/react/structuralSharing.test.ts
Comment (KATT) [inline packages/tests/server/react/structuralSharing.test.ts:1]: Not a blocker, but if you're doing unit tests, I think you can have this right next to the file you're testing rather than in the testing package

Comment (KATT) [reply packages/tests/server/react/structuralSharing.test.ts:1]: Would also like to see some actual integration tests that leverages this - see `useQuery.test.ts` etc

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 3]
```diff
@@ -182,6 +183,7 @@ export function createRootHooks<
       void prefetchQuery(queryKey, opts as any);
     }
     const ssrOpts = useSSRQueryOptionsIfNeeded(queryKey, {
+      structuralSharing: replaceEqualDeep,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:186]: This should be opt-in based on some setting or grabbed from context or whatever

**Review by KATT — DISMISSED** [review]
Needs

- A test that leverages the structural sharing in a test that renders a react app
- Removal of dependency
- Making it opt-in somehow (I'm guessing this is something you pass to the react query proivder?)

File: packages/react-query/src/utils/structuralSharing.ts (hunk 1 above)
Comment (aldebout) [reply packages/react-query/src/utils/structuralSharing.ts]: Done, I copy-pasted the code.

File: www/docs/client/react/setup.mdx [hunk 4]
```diff
@@ -97,6 +97,25 @@ export function App() {
 The reason for using `useState` in the creation of the `queryClient` and the `TRPCClient`, as opposed to declaring them outside of the component, is to ensure that each request gets a unique client when using SSR. If you use client side rendering then you can move them if you wish.
 :::
 
+:::tip
+If you're using a transformer like `superjson`, you probably want React Query to be able to use [structural sharing](https://tanstack.com/query/latest/docs/framework/react/guides/render-optimizations#structural-sharing) for non-JSON types like Date, Set and Map. You can do this by setting the `structuralSharing` option to `replaceEqualDeep` in the `queryClient` options.
```
Comment (KATT) [inline www/docs/client/react/setup.mdx:101]: Add a blurb _why_ you'd want structural sharing 🙏

File: packages/react-query/src/utils/structuralSharing.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
+/**
+ * Taken from @tanstack/query-core utils.ts
+ * Modified to support Date, Map, Set and NaN comparisons
+ *
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ */
+export function replaceEqualDeep(a: any, b: any): any {
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:8]: Since we have this as the only dependency instead, can we just import it instead?

Comment (aldebout) [reply packages/react-query/src/utils/structuralSharing.ts]: As discussed, I updated the wording.

File: packages/react-query/src/utils/structuralSharing.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
+/**
+ * Inspired by @tanstack/query-core utils.ts
+ * Modified to support Date, Map, Set and NaN comparisons
+ *
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ */
+export function replaceEqualDeep(a: any, b: any): any {
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:8]: I think we should allow custom comparators here too - for instance if you use any custom data types like `Temporal.PlainDate` etc?

Could be a factory function like


Also, I'm not sold on the naming 🤔 🤔

File: packages/react-query/src/utils/structuralSharing.test.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
+import { replaceEqualDeep } from '@trpc/react-query';
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.test.ts:1]:
```suggestion
import { replaceEqualDeep } from './structuralSharing';
```

File: packages/react-query/src/utils/structuralSharing.ts (hunk 6 above)
Comment (aldebout) [reply packages/react-query/src/utils/structuralSharing.ts]: I changed to a factory function + a default comparator, let me know what you think.

I  also renamed but I'm still not sure it's great.

File: packages/react-query/src/utils/structuralSharing.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
+/**
+ * Inspired by @tanstack/query-core utils.ts
+ * Modified to support Date, Map, Set and NaN comparisons
+ *
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ */
+export function createStructuralSharingFunction(
+  equalityFn: (a: any, b: any) => boolean,
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:9]: nit:
```suggestion
  equalityFn: (a: unknown, b: unknown) => boolean,
```

**Comment by KATT** [comment]
Thanks for this, will do proper review later in the week!

File: packages/react-query/src/utils/structuralSharing.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+function isPlainArray(value: unknown) {
+  return Array.isArray(value) && value.length === Object.keys(value).length;
+}
+
+// Copied from: https://github.com/jonschlinkert/is-plain-object
+function isPlainObject(o: any): o is object {
+  if (!hasObjectPrototype(o)) {
+    return false;
+  }
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:107]: We have a much simplified version here:

https://github.com/trpc/trpc/blob/3957a7dc564b14f52ce6ba5078490e73b6b6be98/packages/server/src/unstable-core-do-not-import/utils.ts#L26-L32

Which one is wrong? 😅

File: packages/react-query/src/utils/structuralSharing.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  if (a.valueOf !== Object.prototype.valueOf)
+    return a.valueOf() === b.valueOf();
+  if (a.toString !== Object.prototype.toString)
+    return a.toString() === b.toString();
+  return false;
+}
+
+function isPlainArray(value: unknown) {
+  return Array.isArray(value) && value.length === Object.keys(value).length;
+}
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:101]: How come `isPlainArray` rather than `Array.isArray()`? What's the difference?

File: packages/react-query/src/utils/structuralSharing.ts [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    return b;
+  };
+
+  return structuralSharingFunction;
+}
+
+export const defaultStructuralSharingFunction =
+  createStructuralSharingFunction(isEqual);
+
+function isEqual(a: any, b: any) {
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:86]:
```suggestion
export function defaultIsEqual(a: unknown, b: unknown) {
```

export it so it can be used when making a custom `createStructuralSharingFunction`

File: www/docs/client/react/setup.mdx [hunk 12]
```diff
@@ -97,6 +97,27 @@ export function App() {
 The reason for using `useState` in the creation of the `queryClient` and the `TRPCClient`, as opposed to declaring them outside of the component, is to ensure that each request gets a unique client when using SSR. If you use client side rendering then you can move them if you wish.
 :::
 
+:::tip
+React Query uses a technique called [structural sharing](https://tanstack.com/query/latest/docs/framework/react/guides/render-optimizations#structural-sharing) to ensure that as many references as possible will be kept intact between re-renders.
```
Comment (KATT) [inline www/docs/client/react/setup.mdx:101]: I still want an explanation of why it's important to keep values intact between re-renders

I've personally haven't ever felt the need of this optimization and I refetch aggressively [on any mutation](https://trpc.io/docs/client/react/useUtils#invalidate-full-cache-on-every-mutation)

**Review by KATT — CHANGES_REQUESTED** [review]
- Fails CI, please fix
- Also add a motivation on why I'd want to use this

File: packages/react-query/src/utils/structuralSharing.ts (hunk 9 above)
Comment (aldebout) [reply packages/react-query/src/utils/structuralSharing.ts:107]: This comes for RQ and I believe most of it aims to avoid edge cases like this one https://github.com/TanStack/query/pull/7357.
Because we do `copy = {}` if there's a custom constructor we lose something.

File: packages/react-query/src/utils/structuralSharing.ts (hunk 10 above)
Comment (aldebout) [reply packages/react-query/src/utils/structuralSharing.ts:101]: Similar to the object, I believe this is to avoid edge cases https://github.com/TanStack/query/pull/3669.

File: www/docs/client/react/setup.mdx (hunk 12 above)
Comment (aldebout) [reply www/docs/client/react/setup.mdx:101]: I think there's two ways to look at it:

Say you have a page with an "edit" form that uses RHF `useForm` and fetches the current data client-side. There's multiple ways to do this, but there's a chance you have something like

```tsx
useEffect(() => {reset(data.values)}, [data.values])
```

somewhere.

With structural sharing, even if the query is refetched for some reason (for example, you switch tabs with the default behavior), the identity of `data.values` is stable as long at the data stays the same. Without structural sharing, the identity is not stable, the form is reset and you lose any changes you were in the process of making.
I think there are other ways to tackle this specific use-case, but overall I think it's a nice thing that helps react-query work in a more intuitive way.

The second argument is that structural sharing is already present in RQ by default, for json-compatible structures. Since trpc has a RQ integration, I think it makes sense to also provide this feature for those that want it, keeping behavior similar to vanilla RQ while benefiting from the type safety of trpc.
FWIW, stability for query results that contain `Date` objects has been requested / tried [multiple](https://github.com/TanStack/query/pull/6481) [times](https://github.com/TanStack/query/pull/7151) at RQ.

File: packages/react-query/src/utils/structuralSharing.test.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+test('stable for sets - different order is still equal', async () => {
+  const set = new Set([1, 2, 3]);
+  const oldResult = parse(stringify({ a: set }));
+  const newResult = parse(stringify({ a: new Set([3, 2, 1]) }));
+
+  expect(defaultStructuralSharingFunction(oldResult, newResult)).toBe(
+    oldResult,
+  );
+});
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.test.ts:149]: I am wondering if this is a bug or a feature? Is this how RQ handles arrays?

File: packages/react-query/src/utils/structuralSharing.test.ts [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+      ]),
+    }),
+  );
+
+  expect(defaultStructuralSharingFunction(oldResult, newResult)).toBe(
+    oldResult,
+  );
+});
+
+test('stable for maps - different keys are not equal', async () => {
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.test.ts:112]: Not sure if order should matter or not here?

File: packages/react-query/src/utils/structuralSharing.test.ts [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+    oldResult.b,
+  );
+});
+
+test('sets - different sizes are not equal', async () => {
+  const prev = new Set([1]);
+  const next = new Set([1, 2]);
+
+  expect(defaultStructuralSharingFunction(prev, next)).toEqual(next);
+});
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.test.ts:173]: added this test and fixed the bug

File: packages/react-query/src/utils/structuralSharing.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+ *   return false
+ * });
+ */
+export function createStructuralSharingFunction(
+  customEqualityFunction: (a: unknown, b: unknown) => boolean,
+) {
+  const equalCheck = (a: unknown, b: unknown) => {
+    return customEqualityFunction(a, b) || defaultIsEqual(a, b);
+  };
+  const structuralSharingFunction = (prev: any, next: any): any => {
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:22]: renamed these to `prev` and `next` fyi

File: packages/react-query/src/utils/structuralSharing.ts [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    if (equalCheck(prev, next)) {
+      return prev;
+    }
+
+    const array = isPlainArray(prev) && isPlainArray(next);
+    const plainObject = isPlainObject(prev) && isPlainObject(next);
+    const map = prev instanceof Map && next instanceof Map;
+    const set = prev instanceof Set && next instanceof Set;
+
+    if (array || plainObject || map) {
```
Comment (KATT) [inline packages/react-query/src/utils/structuralSharing.ts:32]: The code within here is.... horribly hard to follow

should we split it up into less DRY operations instead?

**Review by KATT — COMMENTED** [review]
changed some variable names, fixed a bug, and left some comments

**Comment by KATT** [comment]
please re-request review or ping me on discord when you have responded etc!

**Comment by PodkopovP** [comment]
Anything happen with this one?
